api/scholarqa/scholar_qa.py

```diff
@@ -206,6 +212,164 @@ def step_gen_iterative_summary(self, query: str, per_paper_summaries: Dict[str,
+            reqd_ref_df["sentence_alpha"] = reqd_ref_df["sentences"].apply(
+                lambda x: [re.sub(r'[^a-zA-Z]', '', sentence["text"]).lower() for sentence in x])
+            # iterate over the reqd_ref_df and get the snippets for each row from reqd_paper_summaries
+            for row_idx, row in reqd_ref_df.iterrows():
+                ref_str, sentences, sent_alpa = row["reference_string"], row["sentences"], row["sentence_alpha"]
+                mapped_quotes = []
+
+                curr_reqd_quotes = reqd_paper_summaries[ref_str].split("...")
+                curr_reqd_quotes_reg = [re.sub(r'[^a-zA-Z]', '', quote).lower() for quote in curr_reqd_quotes]
+                for idx, (quote, quote_reg) in enumerate(zip(curr_reqd_quotes, curr_reqd_quotes_reg)):
+                    new_quote = quote.strip()
+                    curr_quote_map = {"quote": new_quote, "section_title": "abstract",
+                                      "pdf_hash": "", } if not sentences else dict()
+
+                    shift = 0  # keep track of changes to the quote offsets when the inline citations are modified
+                    for sidx, sentence in enumerate(sentences):
+                        # can lookup exact string now since we prompt the llm to include the citations in the quotes
+                        lookup_idx = sentence["text"].lower().find(quote.lower().strip())
+                        raw_match = lookup_idx >= 0
+                        if not raw_match:
+                            lookup_idx = sent_alpa[sidx].find(quote_reg)
+                        if lookup_idx >= 0:
+                            lookup_end = lookup_idx + len(quote)
+                            curr_quote_map["section_title"] = sentence["section_title"]
+                            curr_quote_map["pdf_hash"] = sentence["pdf_hash"]
+                            curr_quote_map["start"], curr_quote_map["end"] = lookup_idx, lookup_end
+                            curr_quote_map["sentence_offsets"], curr_quote_map["ref_mentions"] = [], []
+                            if sentence.get("sentence_offsets"):
+                                for sidx, soff in enumerate(sentence["sentence_offsets"]):
+                                    # check if the sentence offset is within the range of the quote
+                                    # the sentence can be completely or partially inside the quote
+                                    if (lookup_idx < soff["end"] <= lookup_end) or (
+                                            lookup_idx <= soff["start"] < lookup_end):
+                                        curr_quote_map["sentence_offsets"].append(soff)
+                            if sentence.get("ref_mentions"):
+                                for sref in sentence["ref_mentions"]:
+                                    if sref.get("matchedPaperCorpusId") and lookup_idx <= sref.get(
+                                            "start") and sref.get("end") <= lookup_end:
```

Comment:
Check whether each sentence in the passage partially or completely overlaps with the quote, and add it to the sentence_offsets list